### ISSUE TITLE
TreeList selecting parent with selection recursive true, onselectionchanged e.currentSelectedRowKeys only shows parent no children

### ISSUE THREAD

**Issue by genachka** [DevExpress/DevExtreme#2358]
Using 17.2.4 TreeList and for options have `selection: {mode: 'multiple', showCheckBoxesMode: 'always', recursive: true}'. When I select a parent node that has 3 children, it correctly selects those 3 children and the parent however when using callback `onSelectionChanged: function(e)` and looking at `e.currentSelectedRowKeys` it only contains the row I actually clicked on which is the parent row's index. What about the indexes of the children? How do I know that they were also selected. And of course when I unselect parent, the same happens. I need to know all the children that are selected / unselected.

How can I do that?

**Comment by Alyar666** [comment]
Hello,

Currently, our TreeList returns only explicitly selected nodes. We have plans to improve our API to simplify getting all selected rows.

Use the solution from the [TreeList - How to get all selected nodes in recursive selection mode](https://www.devexpress.com/Support/Center/Example/Details/T590425/treelist-how-to-get-all-selected-nodes-in-recursive-selection-mode) example.

**Comment by genachka** [comment]
@Alyar666 Please reopen this as that's not what I'm asking for. The sample provided shows how to get the selected. I'm specifically referring to the callback `onSelectionChanged` that should provide an array of all the selected / unselected keys when you select / unselect the parent whereas currently it only provides the parent key for `e.currentSelectedRowKeys` / `e.currentDeselectedRowKeys` when using the newly recursive (i.e. `selection: {mode: 'multiple', showCheckBoxesMode: 'always', recursive: true}`)

**Comment by Alyar666** [comment]
Hi,

To the onSelectionChanged event handler come only explicitly selected/unselected keys to make the behavior consistent when a remote and a local data source are used. Currently, you will need to iterate through nodes manually. 

_>> The sample provided shows how to get the selected_

The sample provided shows how to get selected child nodes by parent key. You can get selected/unselected child nodes via the getAllSelectedNodes function in the onSelectionChanged event handler.

Here is a complete sample for your convenience: https://codepen.io/anon/pen/MrpNoz

**Comment by genachka** [comment]
@Alyar666 the sample as a workaround helps thanks. Understand current functionality, am requesting enhancement to it for the recursive to be useful with the onSelectionChanged event. I would think that as an enhancement it would return an array of children and array of arrays for each level child of child, etc.

**Comment by GoshaFighten** [comment]
@genachka 

As @Alyar666 stated, we had plans to improve our API to get all selected node. We did this in the #2492 PR. In version 18.1, you will be able to use the getSelectedRowKeys method to do this:

``` JavaScript
treeListInstance.getSelectedRowKeys("all")
```

As for extending onSelectionChanged event parameters, currently, we don't have plans to do this. But, we implemented the forEachNode method in the #2726 PR so that you don't need to write your own code to iterate through all node children. Instead, you are able to simply pass a callback function to the method to specify what you need to do with a node. You can use this method for your task. So, the onSelectionChanged event handler from Alyar666's sample can be re-written as:

``` JavaScript
onSelectionChanged: function (e) {
    var currentSelectedKeys = [],
        currentDeselectedKeys = [];
    e.component.forEachNode(e.currentSelectedRowKeys.map((key) => e.component.getNodeByKey(key)), node => {
        currentSelectedKeys.push(node.key);
    });
    e.component.forEachNode(e.currentDeselectedRowKeys.map((key) => e.component.getNodeByKey(key)), node => {
        currentDeselectedKeys.push(node.key);
    });
    $("#selectionInfo").html(
        "Current selected keys = " + currentSelectedKeys.join(", ") +
        "<br/> Current deselected keys = " + currentDeselectedKeys.join(", ")
    );
}
```

The forEachNode method will be available in versions 17.2.6 and 18.1.

**Comment by genachka** [comment]
@GoshaFighten thanks for the update, clarification and example! I'll give it a shot in 17.2.5 and am eagerly awaiting to try it out in 18.1!

**Comment by sometimesElen** [comment]
Hi @genachka,
We've published our [18.1.0-sprint.2](https://github.com/DevExpress/DevExtreme/releases/tag/18.1.0-sprint.2) results. As you see, we added the `getSelectedRowKeys` overload. But we changed the modes in #2839 . Use the following code to get all selected TreeList nodes
`treeListInstance.getSelectedRowKeys("includeRecursive")`
